adding backward capability for potrf (Cholesky)

jmxpearson opened the pull request:
- cf #440 
- Implementation based on Iain Murray's note (arXiv 1602.07527)
- Uses the symbolic method, which is easier to code and competitive for
  small arrays (according to experiments in the supplement to that paper)
- Unfortunately, the framework in `test_autograd` fails for the `Function` subclass because the outputs of `Potrf` are not independent, so the Jacobian calculation isn't correct. I implemented a different set of tests based on Murray's [here](https://github.com/imurray/chol-rev/blob/master/python/chol_rev_demo.py)

jmxpearson committed:
adding backward capability for potrf (Cholesky)

- based on Iain Murray's note (arXiv 1602.07527)
- uses the symbolic method, which is easier to code and competitive for
  small arrays

fmassa commented on torch/autograd/_functions/linalg.py:
```diff
@@ -102,3 +102,44 @@ def backward(ctx, grad_output, grad_LU=None):
+        """
+        B = A.tril()
+        for i in range(B.size(0)):
+            B[i, i] = B[i, i] * 0.5
```
Might be more efficient to replace this loop with something like
```python
# need to make sure that mat is in the same device as A
mat = torch.ones(A.size()) * (-0.5) + 1
mat = Variable(mat)
B = B * mat
```

jmxpearson replied:
Makes sense. I didn't have a good intuition for whether the speed or the allocation was the bigger issue.

jmxpearson replied:
What's the idiom for making sure `mat` is on the same device? Something like
```python
mat = A.data.new(torch.eye(A.size(0))) * (-0.5) + 1
```
?

fmassa replied:
So, it turns out that there is not yet a version of `torch.eye` that works directly on the GPU.
I implemented it in https://github.com/pytorch/pytorch/pull/2395 , so once it's merged, you could do something like
```python
mat = torch.eye(A.size(0), out=A.data.new()) * (-0.5) + 1
```

jmxpearson replied:
I'll wait for that to merge and then push.

pavanky replied:
It would probably be easier to do the following ?
```
B = B - 0.5 * torch.diag(torch.diag(B))
```

fmassa replied:
Definitely easier, and just slightly less memory efficient. I'd go for @pavanky solution!

fmassa commented on test/test_potrf.py:
```diff
@@ -0,0 +1,50 @@
+
+from torch.autograd._functions.linalg import Potrf
+
+class TestPotrf(TestCase):
```
Can you skip this test if lapack is not found on the system?

fmassa commented on torch/autograd/_functions/linalg.py:
```diff
@@ -102,3 +102,44 @@ def backward(ctx, grad_output, grad_LU=None):
+        return fact
+
+    @staticmethod
+    def Phi(A):
```
can you make this method lowercase?

fmassa reviewed (changes requested):
Did a first pass on the code, overall looks good but I have a few comments. I didn't check the correctness of the gradients, will wait for jenkins on `test_autograd`.
Thanks for the PR!

jmxpearson commented:
I also fixed the `flake8` warnings that were causing Travis to bail. Have made the changes requested and will push again when we get the bit about devices resolved.

jmxpearson commented:
Good deal.
On Sat, Aug 12, 2017 at 11:18 Francisco Massa <notifications@github.com>
wrote:

> *@fmassa* commented on this pull request.
> ------------------------------
>
> In torch/autograd/_functions/linalg.py
> <https://github.com/pytorch/pytorch/pull/2386#discussion_r132820147>:
>
> > +
> +    @staticmethod
> +    def forward(ctx, a, upper=True):
> +        ctx.upper = upper
> +        fact = torch.potrf(a, upper)
> +        ctx.save_for_backward(fact)
> +        return fact
> +
> +    @staticmethod
> +    def Phi(A):
> +        """
> +        Return lower triangle of A and halve the diagonal.
> +        """
> +        B = A.tril()
> +        for i in range(B.size(0)):
> +            B[i, i] = B[i, i] * 0.5
>
> Definitely easier, and just slightly less memory efficient. I'd go for
> @pavanky <https://github.com/pavanky> solution!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/2386#discussion_r132820147>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADBM52M-3fDW0B18y3j36DTIoz8kUExHks5sXcJGgaJpZM4O0qBo>
> .
>

pavanky commented on torch/autograd/_functions/linalg.py:
```diff
@@ -102,3 +102,44 @@ def backward(ctx, grad_output, grad_LU=None):
+
+        P = Potrf.Phi(torch.mm(L.t(), Lbar))
+        S = torch.gesv(P + P.t(), L.t())[0]
+        S = torch.gesv(S.t(), L.t())[0]
```
Wouldn't it be more efficient to use btriact on `L.t()` and call `btrisolve` twice instead of gesv. ?
http://pytorch.org/docs/master/torch.html#torch.btrisolve

fmassa replied:
Maybe, but `btrifact` and `btrisolve` are not yet implemented for `Variable`s, so it would require the function to be implemented via `@once_differentiable`, limiting it's use to first-order derivatives.

jmxpearson committed:
fixes after review

- fixed flake8 errors
- changed phi function to make sure B allocates on the same device as A

soumith commented:
thanks a lot John!
